src/arizona_template_scanner.erl

```diff
@@ -0,0 +1,548 @@
+%%
+%% %CopyrightBegin%
+%%
+%% Copyright 2024 Arizona Framework <contact@arizonaframe.work>
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%%     http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% %CopyrightEnd%
+%%
+-module(arizona_template_scanner).
+-moduledoc false.
+
+%% API functions.
+-export([scan/1]). -ignore_xref([scan/1]).
+
+%% Types
+-export_type([token/0]).
+-export_type([location/0]).
+-export_type([line/0]).
+-export_type([column/0]).
+-export_type([result/0]).
+-export_type([error_reason/0]).
+
+-type state() :: #{
+    line => line(),
+    column => column(),
+    first_column => non_neg_integer(),
+    position => non_neg_integer()
+}.
+-opaque token() :: {open_tag, location(), binary()}
+                 | {attr_key, location(), binary()}
+                 | {attr_value, location(), {text, binary()} | {expr, binary()}}
+                 | {bool_attr, location(), binary()}
+                 | {close_tag, location(), {void, boolean()}}
+                 | {closing_tag, location(), binary()}
+                 | {text, location(), binary()}
+                 | {expr, location(), binary()}.
+-opaque location() :: {line(), column()}.
+-opaque line() :: non_neg_integer().
+-opaque column() :: non_neg_integer().
+-opaque result() :: {ok, [token()]} | {error, {error_reason(), location()}}.
+-opaque error_reason() :: badexpr
+                        | unexpected_comment
+                        | unexpected_tag_end
+                        | unexpected_expr_end
+                        | unexpected_string_end.
+
+%% --------------------------------------------------------------------
+%% API funtions.
+%% --------------------------------------------------------------------
+
+-spec scan(Bin) -> Result
+    when Bin :: binary(),
+         Result :: result().
+scan(Bin) when is_binary(Bin) ->
+    try
+        {ok, scan(Bin, Bin, 0, new_state())}
+    catch
+        throw:{Reason, State} ->
+            {error, {Reason, location(State)}}
+    end.
+
+%% --------------------------------------------------------------------
+%% Internal funtions.
+%% --------------------------------------------------------------------
+
+-spec scan(Rest, Bin, Len, State) -> Tokens
+    when Rest :: binary(),
+         Bin :: binary(),
+         Len :: non_neg_integer(),
+         State :: state(),
+         Tokens :: [token()].
```

Comment:
If this isn't required, for the time being, I'd remove it. I used to like to have everything spec'ed, but internals, unless they're super convoluted to understand, I won't.

The fact that e.g. `Rest` and `Bin` are spec'ed as binaries is not helpful because I won't make less of a mistake using them. If we could spec the binaries that would be a different story. The other variables' types we can basically infer from just looking at them 😄

Reply:
Removing it I receive 
```
    ┌─ src/arizona_template_scanner.erl:
    │
 34 │  -type state() :: #{
    │   ╰── type state() is unused
```
I'm thinking of keeping it due to this. Otherwise the state will not have a type. 
Any suggestion?

Reply:
Yeah, of you wanna have the state declared (and this is the only place it's used) when we keep it.

Reply:
I'm reopening this conversation because now I see I've used a map for the state. We can safely remove this spec by changing the state to a record.

This change is in [2807425](https://github.com/arizona-framework/arizona/pull/143/commits/2807425cc00adf82b3f678e9bde3a4672c327186)